docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,393 @@
+	ID string `json:id"`
+
+	// type defines the type of IBM Cloud resource.
+	Type VPCResourceType `json:type"`
+}
+
+type VPCResourceType string
+
+const (
+	VPCResourceTypeCOS               VPCResourceType = VPCResourceType("cos")
+	VPCResourceTypePublicGateway     VPCResourceType = VPCResourceType("publicgateway")
+	VPCResourceTypeSecurityGroup     VPCResourceType = VPCResourceType("securitygroup")
+	VPCResourceTypeSecurityGroupRule VPCResourceType = VPCResourceType("securitygrouprule")
+	VPCResourceTypeSubnet            VPCResourceType = VPCResourceType("subnet")
+	VPCResourceTypeVPC               VPCResourceType = VPCResourceType("vpc")
+)
+```
+
+
+#### Share existing common resource definitions
+Rather than duplicate code and also make references to existing resources
+complex, we also propose relocating existing resource references into a
+common file. This prevents the duplication of code in PowerVS and VPC
+definitions, but also removes the direct dependence on PowerVS definitions,
+placing them into a common location (preventing unexpected dependencies).
+
+Relocate the following resource definitions in IBMPowerVSCluster
+```go
+// VPCResourceReference is a reference to a specific VPC resource by ID or Name
+// Only one of ID or Name may be specified. Specifying more than one will result in
+// a validation error.
+type VPCResourceReference struct {
+	// ID of resource
+	// +kubebuilder:validation:MinLength=1
+	// +optional
+	ID *string `json:"id,omitempty"`
+
+	// Name of resource
+	// +kubebuilder:validation:MinLength=1
+	// +optional
+	Name *string `json:"name,omitempty"`
+
+	// IBM Cloud VPC region
+	Region *string `json:"region,omitempty"`
+}
+
+// CosInstance represents IBM Cloud COS instance.
+type CosInstance struct {
+	// PresignedURLDuration defines the duration for which presigned URLs are valid.
+	//
+	// This is used to generate presigned URLs for S3 Bucket objects, which are used by
+	// control-plane and worker nodes to fetch bootstrap data.
+	//
+	// When enabled, the IAM instance profiles specified are not used.
+	// +optional
+	PresignedURLDuration *metav1.Duration `json:"presignedURLDuration,omitempty"`
+
+	// Name defines name of IBM cloud COS instance to be created.
+	// +kubebuilder:validation:MinLength:=3
+	// +kubebuilder:validation:MaxLength:=63
+	// +kubebuilder:validation:Pattern=`^[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$`
+	Name string `json:"name,omitempty"`
+
+	// bucketName is IBM cloud COS bucket name
+	BucketName string `json:"bucketName,omitempty"`
+
+	// bucketRegion is IBM cloud COS bucket region
+	BucketRegion string `json:"bucketRegion,omitempty"`
+}
+```
+
+To a shared location, such as [types.go](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/api/v1beta2/types.go).
+
+We also are adding additional resources and definitions that could be shared,
+such as the following:
+```go
+// PortRange represents a range of ports, minimum to maximum.
+type PortRange struct {
```

Comment:
I dont have much idea about the security group section, may be will try to understand better before taking a look,